Homework six

Hunkizzz opened the pull request.

Hunkizzz committed:
homework-six

Hunkizzz committed:
homework-six

Hunkizzz committed:
homework-six

Hunkizzz committed:
homework-six

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/dao/AuthorDaoImpl.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.List;
+import java.util.Optional;
+
+@Repository
```
Что тут мы используем от аннотации @repository? Она несколько больше чем @Component или @Service. Если она появляется -то мы ожидаем использование её особенностей

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/dao/BookDaoImpl.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    @Override
+    public void updateNameById(long id, String name) {
```
Задача относится к бизнес-слою. Небольшое упрощение обновления почти не скажется на производительности, но усложнит код. по идеологии дао слоя - тут максимально простые и переиспользуемые операции, всё остальное - уже бизнес-логика

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/dao/BookDaoImpl.java:
```diff
@@ -0,0 +1,100 @@
+
+    @Override
+    public void deleteById(long id) {
+        Query query = entityManager.createQuery("delete " +
```
А чем em.remove плох? Такой подход достаточно опасен, потому что работает в обход кэша и сущность может там остаться, а из базы уже удалили

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/dao/CommentDaoImpl.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Override
+    public void deleteByBookId(long id) {
+        Query query = entityManager.createQuery("delete " +
```
Дублирование кода, от deleteById почти не отличается

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Override
+    @Transactional(propagation = Propagation.SUPPORTS, readOnly = true)
```
Какая тут польза от транзакции? Пока я тут вижу только дополнительные накладные расходы

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/model/Book.java:
```diff
@@ -0,0 +1,58 @@
+    private Long id;
+    @Column(name = "title")
+    private String title;
+    @ManyToOne(cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REFRESH}, fetch = FetchType.EAGER)
```
не надо повторять значения по умолчанию. это смущает при чтении и усложняет сопровождение

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/model/Book.java:
```diff
@@ -0,0 +1,58 @@
+        this.title = title;
+    }
+
+    public Book(String title, Genre genre, Author author) {
```
4 конструтора у класса из 4 полей.. Усложнение, можно уже билдером обойтись.

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/model/Book.java:
```diff
@@ -0,0 +1,58 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(name = "title")
```
Стоит делать пустую строчку между объявлениями полей, это упросит чтение кода, будет четко видно где отдельное поле объявляется

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    @Transactional(propagation = Propagation.SUPPORTS, readOnly = true)
+    public List<Comment> findAllCommentsByAuthorId(long id) {
```
Не очень понятна суть данного запроса с точки зрения бизнес-задачи, но пускай. 
Давайте ещё добавим получение комментариев по книге. при этом не надо писать для этого запрос специальный в dao слое, доработаем связи и используем их

Hunkizzz committed:
homework-six fix

Hunkizzz committed:
homework-six fix

Hunkizzz committed:
homework-six fix

Hunkizzz committed:
homework-seven

Hunkizzz committed:
homework-seven

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/dao/BookDaoImpl.java:
```diff
@@ -0,0 +1,86 @@
+import java.util.List;
+import java.util.Optional;
+
+@Repository
```
не везде дочищено repository

Hunkizzz replied:
Заменил на component

Hunkizzz committed:
homework-six change @repository on component

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/dao/CommentDaoImpl.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Override
+    public void updateTextById(long id, String text) {
+        Query query = entityManager.createQuery("update Comment comment " +
```
задача уровня бизнес-слоя. Я не каждый момент подсвечивал комментрием, отмечаю типовые вопросы с которыми стоит поработать.

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/dao/CommentDaoImpl.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @Override
+    public List<Comment> findByBookId(long id) {
```
Давайте это сделаем не в dao слое, а на уровне бизнес-логики, поработаем по связями через ORM (book.getComments)

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/controller/ShellController.java:
```diff
@@ -0,0 +1,134 @@
+    public void showAllBooksByAuthorId() {
+        ioService.write("Введите Id автора для отображения списка его книг");
+        long id = ioService.readInt();
+        List<Book> books = bookService.findAllBooksByAuthorId(id);
```
Вот это всё задача уровня бизнес-слоя. Тут только запросить информацию, дальше передать в сервис (1 вызов) и отобразить результат. А знать как бизнес-слои связанны между собой - не задача слоя представления

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    @Transactional
+    public void addNewComment() {
+        ioService.write("Введите id книги для добавления комментария");
```
А тут обратное проникновение, задачи UI в бизнес слое. Идеологически в shell мы запросили нужную информацию (id книги и текст коммента), потом их передали в addNewComment и далее уже на основании результата Shell слой отписал итог клиенту

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    @Transactional
+    public void updateTextById(long id, String text) {
+        commentDao.updateTextById(id, text);
```
Тут давайте попробуем обновить без отдельного метода в dao, используя возможности ORM

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/model/Book.java:
```diff
@@ -0,0 +1,62 @@
+    @JoinColumn(name = "genre_id")
+    private Genre genre;
+
+    @OneToMany(fetch = FetchType.EAGER
```
Тут ни в коем случае не EAGER, иначе мы будем постоянно за тянуть всю базу. Когда мы смотрим список книг - то нам не нужны комментарии (для примера)

Hunkizzz committed:
homework-six

remove ioService from another services.
Change Update queries with ORM functions.

Hunkizzz committed:
homework-seven

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/model/Book.java:
```diff
@@ -0,0 +1,58 @@
+    private String title;
+    @ManyToOne(cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REFRESH})
+    @JoinColumn(name = "author_id")
+    private Author author;
```
Стоит разделять объявления пустой строкой всё же, читаться будет лучше

kataus commented on homework-six/src/main/java/ru/otus/education/jpalibraryapp/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,65 @@
+    @Override
+    public List<Comment> findByBookId(long id) {
+        Book book = bookService.findById(id);
+        return book.getComments();
```
Сейчас тут будет LazyInitException

Hunkizzz committed:
homework-six

Hunkizzz committed:
Merge pull request #9 from Hunkizzz/homework-seven

Homework seven
